pkg/asyncui/asyncui.go

```diff
@@ -0,0 +1,143 @@
+package asyncui
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kyma-incubator/hydroform/parallel-install/pkg/components"
+	"github.com/kyma-incubator/hydroform/parallel-install/pkg/deployment"
+	"github.com/kyma-project/cli/pkg/step"
+)
+
+// StepFactory is a factory used to generate a step in the UI.
+type StepFactory interface {
+	NewStep(msg string) step.Step
+}
+
+// End-user messages
+const (
+	deployPrerequisitesPhaseMsg   string = "Deploying pre-requisites"
+	undeployPrerequisitesPhaseMsg string = "Undeploying pre-requisites"
+	deployComponentsPhaseMsg      string = "Deploying Kyma"
+	undeployComponentsPhaseMsg    string = "Undeploying Kyma"
+	deployComponentMsg            string = "Deploying component '%s'"
+)
+
+// AsyncUI renders the CLI ui based on receiving events
+type AsyncUI struct {
+	// used to create UI steps
+	StepFactory StepFactory
+	// channel provided by caller to retrieve errors
+	Errors chan<- error
+	// processing context
+	context context.Context
+	cancel  context.CancelFunc
+	// channel to retreive update events
+	updates chan deployment.ProcessUpdate
+	// internal state
+	running bool
+}
+
+// Start renders the CLI UI and provides the channel for receiving events
+func (ui *AsyncUI) Start() (chan deployment.ProcessUpdate, error) {
+	if ui.running {
+		return nil, fmt.Errorf("Duplicate call of start method detected")
+	}
+	ui.running = true
+
+	// process async process updates
+	ui.updates = make(chan deployment.ProcessUpdate)
+	// initialize processing context
+	ui.context, ui.cancel = context.WithCancel(context.Background())
+
+	go func() {
+		defer ui.cancel()
+		ongoingSteps := make(map[deployment.InstallationPhase]step.Step)
+		for procUpdateEvent := range ui.updates {
+			switch procUpdateEvent.Event {
+			case deployment.ProcessRunning:
+				// Component related update event (components have no ProcessStart/ProcessStop event)
+				if procUpdateEvent.Component.Name != "" {
+					err := ui.renderStopEvent(procUpdateEvent, &ongoingSteps)
+					ui.sendError(err)
+				}
+				continue
+			case deployment.ProcessStart:
+				err := ui.renderStartEvent(procUpdateEvent, &ongoingSteps)
+				ui.sendError(err)
+			default:
+				err := ui.renderStopEvent(procUpdateEvent, &ongoingSteps)
+				ui.sendError(err)
+			}
+		}
+	}()
+
+	return ui.updates, nil
+}
+
+func (ui *AsyncUI) sendError(err error) {
+	if err != nil && ui.Errors != nil {
+		ui.Errors <- err
+	}
+}
```

Comment:
As far as I see, this error delivery mechanism is not yet used (and also not tested). Maybe we can spit until we actually need it. What would you say?